lib/compilers/coccinelle.ts

```diff
@@ -0,0 +1,281 @@
+        filters?: ParseFiltersAndOutputOptions,
+    ): Promise<CompilationResult> {
+        // here to extract contents of source and semantic patch
+        if (!execOptions) {
+            execOptions = this.getDefaultExecOptions();
+        }
+
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = path.dirname(inputFilename);
+        }
+
+        const spatchFilename = path.join(path.dirname(inputFilename), this.spatchBaseFilename);
+        let cFileContents = '';
+        let pFileContents = '';
+        let toc = true; // toc: Target Language or Coccinelle Language
+        const file = await open(inputFilename);
+        for await (const line of file.readLines()) {
+            // this separates the C/C++ portion from the SmPL
+            const ifdefCocciRegex = /^#ifdef.*COCCINELLE.*/;
+            const endifCocciRegex = /^#endif.*COCCINELLE.*/;
+
+            if (ifdefCocciRegex.test(line) && toc) {
+                toc = !toc;
+            } else {
+                if (!endifCocciRegex.test(line)) {
+                    if (toc) cFileContents += line + '\n';
+                    else pFileContents += line + '\n';
+                }
+            }
+        }
+        file.close();
+
+        // we overwrite the C/C++ + SmPL with extracted C/C++ sources
+        fs.writeFileSync(inputFilename, cFileContents);
+        // we save the extracted SmPL sources
+        fs.writeFileSync(spatchFilename, pFileContents);
+
+        // check that the SmPL source does not #include anything, via CE rules
+        let cocciSourceError = this.checkSource(pFileContents);
+        if (cocciSourceError) throw cocciSourceError;
+        // use own rules, too
+        cocciSourceError = this.checkCocciSource(pFileContents);
+        if (cocciSourceError) throw cocciSourceError;
+
+        const result = await this.exec(compiler, options, execOptions);
+
+        if (this.joinSpatchStdinAndStderr && result.code === 0 && !result.timedOut) {
+            result.stderr += result.stdout;
+            result.stdout = result.stderr;
+            result.stderr = '';
+        }
+        return {
+            ...this.transformToCompilationResult(result, inputFilename),
+            languageId: this.getCompilerResultLanguageId(filters),
+            instructionSet: this.getInstructionSetFromCompilerArgs(options),
+        };
+    }
+
+    override getOutputFilename(dirPath: string, outputFilebase: string, key?: CacheKey): string {
+        return path.join(dirPath, `${outputFilebase}${this.lang.extensions[0]}`);
+    }
+
+    override postCompilationPreCacheHook(result: CompilationResult): CompilationResult {
+        if (result.code === 0) {
+            result.languageId = 'c'; // we produce C
+        }
+        return result;
+    }
+
+    override filterUserOptions(args: string[]) {
+        const permittedOptions = new Set(['--debug', '--verbose', '--quiet']); // preliminary and in need of a mention somewhere
```

Comment:
I'm not quite sure what this means here. Why do you need to filter the user options here in particular? Is there something you're trying to prevent? (the user option filter is not a place to add security related issues as it's often easy to work around).

Reply:
I want to prevent the user from using options which may access arbitrary files on the server running Compiler Explorer.